docs/src/modules/components/ThemeContext.js

```diff
@@ -0,0 +1,113 @@
+import { createMuiTheme } from '@material-ui/core/styles';
+import blue from '@material-ui/core/colors/blue';
+import pink from '@material-ui/core/colors/pink';
+import { darken } from '@material-ui/core/styles/colorManipulator';
+import { getCookie } from 'docs/src/modules/utils/helpers';
+import { lightTheme, darkTheme, setPrismTheme } from 'docs/src/modules/components/prism';
+
+export const themeInitialOptions = {
+  direction: 'ltr',
+  paletteType: 'light',
+  paletteColors: {
+    primary: blue,
+    secondary: {
+      // Darken so we reach the AA contrast ratio level.
+      main: darken(pink.A400, 0.08),
+    },
+  },
+};
+
+export const DispatchContext = React.createContext(() => {
+  throw new Error('Forgot to wrap component in ThemeContext.Provider');
+});
+
+export function Provider(props) {
+  const { children } = props;
+
+  const [themeOptions, dispatch] = React.useReducer((state, action) => {
+    switch (action.type) {
+      case 'RESET_COLORS':
+        return {
+          ...state,
+          paletteColors: themeInitialOptions.paletteColors,
+        };
+      case 'CHANGE':
+        return {
+          paletteType: action.payload.paletteType || state.paletteType,
+          direction: action.payload.direction || state.direction,
+          paletteColors: action.payload.paletteColors || state.paletteColors,
+        };
+      default:
+        throw new Error(`Unrecognized type ${action.type}`);
+    }
+  }, themeInitialOptions);
+  const { direction, paletteColors, paletteType } = themeOptions;
+
+  React.useEffect(() => {
+    setPrismTheme(paletteType === 'light' ? lightTheme : darkTheme);
+  }, [paletteType]);
+
+  const theme = React.useMemo(() => {
+    const nextTheme = createMuiTheme({
+      direction,
+      nprogress: {
+        color: paletteType === 'light' ? '#000' : '#fff',
+      },
+      palette: {
+        ...paletteColors,
+        type: paletteType,
+        background: {
+          default: paletteType === 'light' ? '#fff' : '#303030',
+        },
+      },
+    });
+
+    nextTheme.palette.background.level1 =
+      paletteType === 'light' ? nextTheme.palette.grey[100] : nextTheme.palette.grey[900];
+
+    nextTheme.palette.background.level0 =
+      paletteType === 'light' ? nextTheme.palette.grey[50] : nextTheme.palette.grey[900];
+
+    return nextTheme;
+  }, [direction, paletteColors, paletteType]);
+
+  React.useEffect(() => {
+    // Expose the theme as a global variable so people can play with it.
+    if (process.browser) {
+      window.theme = theme;
+    }
+  }, [theme]);
+
+  React.useEffect(() => {
+    if (process.browser) {
+      const nextPaletteColors = JSON.parse(getCookie('paletteColors') || 'null');
+      const nextPaletteType = getCookie('paletteType');
+
+      dispatch({
+        type: 'CHANGE',
+        payload: { paletteColors: nextPaletteColors, paletteType: nextPaletteType },
+      });
+    }
+  }, []);
+
+  return (
+    <ThemeProvider theme={theme}>
```

Comment:
I have already mentioned this point, but I think that it's important to stress it. (This is a way broader point than our 1 redux store vs x custom contexts issue here. I'm happy to move forward with this ThemeProvider, We can give it a try 👌).

In my perspective, the more **monolithic** the code base is, the most productive we are. In the backend world, a monolith is simpler to orchestrate than micro-services. I think that the default should be to have everything grouped but to move away from it (isolation) when, and only when, **we hit a wall**.

Reply:
micro-services have nothing to do with it. What exactly were you grouping before? It definitely wasn't anything related to theming. That was cluttered across the codebase.

Reply:
Micro-services was an example of how allowing more degrees of freedom can hit you hard and more importantly, often out weights the benefits. This issue was only a pretext to share my perspective on the problem. I know from my previous experience at Onepixel and Doctolib that this is a point people often disagree on.